services/mlmodel/mlmodel.go

```diff
@@ -0,0 +1,255 @@
+	pbmm := &servicepb.Metadata{
+		Name:        mm.ModelName,
+		Type:        mm.ModelType,
+		Description: mm.ModelDescription,
+	}
+	inputInfo := make([]*servicepb.TensorInfo, 0, len(mm.Inputs))
+	for _, inp := range mm.Inputs {
+		inproto, err := inp.ToProto()
+		if err != nil {
+			return nil, err
+		}
+		inputInfo = append(inputInfo, inproto)
+	}
+	pbmm.InputInfo = inputInfo
+	outputInfo := make([]*servicepb.TensorInfo, 0, len(mm.Outputs))
+	for _, outp := range mm.Outputs {
+		outproto, err := outp.ToProto()
+		if err != nil {
+			return nil, err
+		}
+		outputInfo = append(outputInfo, outproto)
+	}
+	pbmm.OutputInfo = outputInfo
+	return pbmm, nil
+}
+
+// TensorInfo contains all the information necessary to build a struct from the input and output maps.
+// it describes the name of the output field, what data type it has, and how many dimensions the
+// array/tensor will have. AssociatedFiles points to where more information is located, e.g. in case the ints
+// within the array/tensor need to be converted into a string.
+type TensorInfo struct {
+	Name            string // e.g. bounding_boxes
+	Description     string
+	DataType        string // e.g. uint8, float32, int
+	NDim            int    // number of dimensions in the array
+	AssociatedFiles []File
+	Extra           map[string]interface{}
+}
+
+// ToProto turns the TensorInfo struct into a protobuf message.
+func (tf TensorInfo) ToProto() (*servicepb.TensorInfo, error) {
+	pbtf := &servicepb.TensorInfo{
+		Name:        tf.Name,
+		Description: tf.Description,
+		DataType:    tf.DataType,
+		NDim:        int32(tf.NDim),
+	}
+	associatedFiles := make([]*servicepb.File, 0, len(tf.AssociatedFiles))
+	for _, af := range tf.AssociatedFiles {
+		afproto, err := af.ToProto()
+		if err != nil {
+			return nil, err
+		}
+		associatedFiles = append(associatedFiles, afproto)
+	}
+	pbtf.AssociatedFiles = associatedFiles
+	extra, err := vprotoutils.StructToStructPb(tf.Extra)
+	if err != nil {
+		return nil, err
+	}
+	pbtf.Extra = extra
+	return pbtf, nil
+}
+
+// File contains information about how to interpret the numbers within the tensor/array. The label type
+// describes how to read the tensor in order to successfully label the numbers.
+type File struct {
+	Name        string // e.g. category_labels.txt
+	Description string
+	LabelType   LabelType // TENSOR_VALUE, or TENSOR_AXIS
+}
+
+// ToProto turns the File struct into a protobuf message.
+func (f File) ToProto() (*servicepb.File, error) {
+	pbf := &servicepb.File{
+		Name:        f.Name,
+		Description: f.Description,
+	}
+	switch f.LabelType {
+	case LabelTypeUnspecified:
+		pbf.LabelType = servicepb.LabelType_LABEL_TYPE_UNSPECIFIED
+	case LabelTypeTensorValue:
+		pbf.LabelType = servicepb.LabelType_LABEL_TYPE_TENSOR_VALUE
+	case LabelTypeTensorAxis:
+		pbf.LabelType = servicepb.LabelType_LABEL_TYPE_TENSOR_AXIS
+	default:
+		return nil, errors.Errorf("do not know about ML Model associated file LabelType %q", f.LabelType)
```

Comment:
I actually don't think we should error here. Looks like that'll trickle up and prevent the whole Metadata.toProto() from vibing just because of this lack of info.  Maybe just warn if possible and set to unspecified so we can try to get what we can outta Metadata()

Reply:
done